# **Imports**

In [2]:
# Preprocessing
import zipfile as zf
import pandas as pd
import numpy as np
import os
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.decomposition import PCA
# Add your libraries here

# **Read in the data**

In [20]:
# Use the current working directory
current_dir = os.getcwd()

# Check if it's in the local/Kaggle environment to declare path variables
if current_dir == "/kaggle/working":
    file_test_path = "/kaggle/input/dog-breeds-recognition/dog-breeds-recognition/testing.csv"
    file_train_path = "/kaggle/input/dog-breeds-recognition/dog-breeds-recognition/training.csv"
else:
    file_test_path = "/content/dog-breeds-recognition/testing.csv"
    file_train_path = "/content/dog-breeds-recognition/training.csv"
    
print(f"Test set: {file_test_path}")
print(f"Train set: {file_train_path}")

Test set: /kaggle/input/dog-breeds-recognition/dog-breeds-recognition/testing.csv
Train set: /kaggle/input/dog-breeds-recognition/dog-breeds-recognition/training.csv


In [6]:
# read in train
df_train = pd.read_csv(file_train_path, header=None)
print(f"Are there any null values in Train: {df_train.isnull().values.any()}")
df_train

Are there any null values in Train: False


0    1         2    \
0                 n02085620-Chihuahua\n02085620_5927.jpg    1  3.508880   
1                 n02085620-Chihuahua\n02085620_4441.jpg    1  0.447207   
2                 n02085620-Chihuahua\n02085620_1502.jpg    1  1.730776   
3                 n02085620-Chihuahua\n02085620_1916.jpg    1  1.986778   
4                n02085620-Chihuahua\n02085620_13151.jpg    1  0.000000   
...                                                  ...  ...       ...   
11995  n02116738-African_hunting_dog\n02116738_10614.jpg  120  0.051192   
11996   n02116738-African_hunting_dog\n02116738_9282.jpg  120  2.393783   
11997   n02116738-African_hunting_dog\n02116738_6754.jpg  120  0.082882   
11998   n02116738-African_hunting_dog\n02116738_9333.jpg  120  0.027123   
11999   n02116738-African_hunting_dog\n02116738_2503.jpg  120  0.374718   

            3         4         5         6         7         8         9    \
0      0.928564  0.298451  0.202423  0.273040  0.073741  0.260721  2.454843   
1      0.152954  0.214087  1.132086  0.984579  0.352944  0.616292  1.692439   
2      0.405669  0.187414  0.365856  0.512063  0.772889  0.267891  0.160474   
3      0.475547  0.114825  0.263515  0.743351  0.053917  0.293086  0.466959   
4      0.993176  0.362772  0.117868  0.257813  0.817096  0.777969  0.930595   
...         ...       ...       ...       ...       ...       ...       ...   
11995  0.263571  1.953465  0.144953  1.881053  1.025135  1.309460  0.026804   
11996  0.583523  0.070011  2.168356  2.056204  0.418436  0.592864  2.390873   
11997  0.210209  1.683243  0.442680  2.221987  1.028761  2.321181  0.150616   
11998  0.675996  0.353758  0.421884  0.610241  0.311903  1.028679  0.632097   
11999  0.627293  0.173531  0.000000  1.390403  0.759935  2.652178  0.130795   

       ...       504       505       506       507       508       509  \
0      ...  0.929986  0.959449  1.033072  0.104633  1.102810  0.058879   
1      ...  1.842100  0.193077  0.103465  1.228296  2.554505  0.303815   
2      ...  2.033174  0.272994  0.217314  1.633859  1.064902  0.221387   
3      ...  1.581877  0.216979  0.313218  2.389633  1.645630  0.348484   
4      ...  1.301286  0.631204  0.194294  0.170246  2.551018  0.000000   
...    ...       ...       ...       ...       ...       ...       ...   
11995  ...  0.804875  0.278833  0.003926  0.465810  0.050890  1.009473   
11996  ...  0.225061  0.229859  0.935760  0.655962  0.289180  0.623364   
11997  ...  0.237797  0.049576  0.205640  0.424587  0.558180  0.732412   
11998  ...  1.706490  0.396890  0.092548  1.995632  0.000000  1.191216   
11999  ...  0.097647  0.244688  0.047552  0.540931  0.298368  0.637011   

            510       511       512       513  
0      0.508010  0.386740  0.182575  0.764517  
1      0.544647  0.715426  0.916348  0.240079  
2      1.036172  0.591452  0.733662  1.475364  
3      0.936218  0.733054  0.740681  1.340304  
4      3.035839  0.058260  1.350222  3.290168  
...         ...       ...       ...       ...  
11995  1.667273  1.975637  0.620393  0.057779  
11996  1.158500  1.949258  1.581664  0.121391  
11997  1.422601  2.745240  0.044735  0.282376  
11998  0.196342  0.957877  0.736384  0.034472  
11999  1.254008  1.859550  0.309276  0.187103  

[12000 rows x 514 columns]

In [7]:
# read in test
df_test = pd.read_csv(file_test_path, header=None)
print(f"Are there any null values in Test: {df_test.isnull().values.any()}")
df_test

Are there any null values in Test: False


0    1         2    \
0               n02085620-Chihuahua\n02085620_2650.jpg    1  2.750876   
1               n02085620-Chihuahua\n02085620_4919.jpg    1  3.723587   
2               n02085620-Chihuahua\n02085620_1765.jpg    1  0.971007   
3               n02085620-Chihuahua\n02085620_3006.jpg    1  1.874442   
4               n02085620-Chihuahua\n02085620_1492.jpg    1  0.254995   
...                                                ...  ...       ...   
8575  n02116738-African_hunting_dog\n02116738_4991.jpg  120  0.068341   
8576  n02116738-African_hunting_dog\n02116738_3024.jpg  120  0.012424   
8577  n02116738-African_hunting_dog\n02116738_3635.jpg  120  0.090567   
8578  n02116738-African_hunting_dog\n02116738_2988.jpg  120  0.249121   
8579  n02116738-African_hunting_dog\n02116738_6330.jpg  120  0.006490   

           3         4         5         6         7         8         9    \
0     0.402823  0.490048  0.088944  0.067087  0.097192  0.188324  0.554036   
1     1.346295  0.413855  1.559932  2.922223  0.046513  0.065089  0.748078   
2     0.447430  0.136485  0.470510  0.371954  0.112476  0.265642  0.300397   
3     1.340326  0.430788  1.061602  0.018172  0.659020  0.024218  0.541496   
4     1.975686  0.795596  0.792883  0.491056  1.389314  0.066709  0.368587   
...        ...       ...       ...       ...       ...       ...       ...   
8575  0.824908  0.193437  0.906468  2.191380  0.567653  2.182863  0.541500   
8576  0.074204  1.992666  0.076113  0.573071  1.246892  0.586113  0.114273   
8577  0.394184  1.207222  0.123069  2.618800  0.823760  3.099759  0.373843   
8578  0.176359  2.857338  0.408916  0.582186  1.393884  2.366030  0.707135   
8579  0.197918  0.512689  0.059964  0.486442  0.658949  3.135226  1.372569   

      ...       504       505       506       507       508       509  \
0     ...  1.009560  0.560529  0.000568  1.048601  1.888090  0.000000   
1     ...  0.800828  0.310406  0.141535  0.613455  1.598750  0.285342   
2     ...  0.944592  1.212590  0.189738  1.528287  2.109546  0.329096   
3     ...  2.426595  0.921312  0.569086  0.939952  1.300098  1.269613   
4     ...  1.439100  1.203172  0.787512  0.178735  0.667279  0.644420   
...   ...       ...       ...       ...       ...       ...       ...   
8575  ...  0.270837  0.148257  0.538312  0.221652  1.261061  0.393819   
8576  ...  0.544384  0.052004  0.279680  0.214920  1.575158  1.586292   
8577  ...  0.082895  0.005232  0.012785  0.310990  0.028843  1.175330   
8578  ...  0.592481  0.396963  0.142146  0.550736  0.031739  0.102475   
8579  ...  0.280469  0.263225  0.788225  1.298100  0.071682  0.007359   

           510       511       512       513  
0     0.072166  0.522479  0.072563  0.040364  
1     2.139037  1.185468  2.046112  1.761059  
2     1.496220  1.040246  0.000382  1.097506  
3     0.131599  0.544578  0.598401  0.251547  
4     1.120357  1.260047  0.071751  1.355075  
...        ...       ...       ...       ...  
8575  0.931265  1.368547  0.725584  1.802775  
8576  1.686255  0.970767  1.198236  0.281134  
8577  0.576579  1.986010  0.086537  0.805687  
8578  2.250938  0.885802  1.031099  0.105213  
8579  1.349891  2.497369  0.263831  0.009741  

[8580 rows x 514 columns]

# **Reformat Dataframe**
* The dataset doesn't come with column names, lets add some

In [9]:
def reformat_df(df):
  df = df.rename(columns={0:"image_path", 1:"target"})
  for header in df.columns[2:]:
    new_header = f"feature_{header-1}"
    df.rename(columns={header:new_header}, inplace=True)
  df = df.set_index("image_path")
  return df

In [10]:
df_train = reformat_df(df_train)
df_train

target  feature_1  \
image_path                                                             
n02085620-Chihuahua\n02085620_5927.jpg                  1   3.508880   
n02085620-Chihuahua\n02085620_4441.jpg                  1   0.447207   
n02085620-Chihuahua\n02085620_1502.jpg                  1   1.730776   
n02085620-Chihuahua\n02085620_1916.jpg                  1   1.986778   
n02085620-Chihuahua\n02085620_13151.jpg                 1   0.000000   
...                                                   ...        ...   
n02116738-African_hunting_dog\n02116738_10614.jpg     120   0.051192   
n02116738-African_hunting_dog\n02116738_9282.jpg      120   2.393783   
n02116738-African_hunting_dog\n02116738_6754.jpg      120   0.082882   
n02116738-African_hunting_dog\n02116738_9333.jpg      120   0.027123   
n02116738-African_hunting_dog\n02116738_2503.jpg      120   0.374718   

                                                   feature_2  feature_3  \
image_path                                                                
n02085620-Chihuahua\n02085620_5927.jpg              0.928564   0.298451   
n02085620-Chihuahua\n02085620_4441.jpg              0.152954   0.214087   
n02085620-Chihuahua\n02085620_1502.jpg              0.405669   0.187414   
n02085620-Chihuahua\n02085620_1916.jpg              0.475547   0.114825   
n02085620-Chihuahua\n02085620_13151.jpg             0.993176   0.362772   
...                                                      ...        ...   
n02116738-African_hunting_dog\n02116738_10614.jpg   0.263571   1.953465   
n02116738-African_hunting_dog\n02116738_9282.jpg    0.583523   0.070011   
n02116738-African_hunting_dog\n02116738_6754.jpg    0.210209   1.683243   
n02116738-African_hunting_dog\n02116738_9333.jpg    0.675996   0.353758   
n02116738-African_hunting_dog\n02116738_2503.jpg    0.627293   0.173531   

                                                   feature_4  feature_5  \
image_path                                                                
n02085620-Chihuahua\n02085620_5927.jpg              0.202423   0.273040   
n02085620-Chihuahua\n02085620_4441.jpg              1.132086   0.984579   
n02085620-Chihuahua\n02085620_1502.jpg              0.365856   0.512063   
n02085620-Chihuahua\n02085620_1916.jpg              0.263515   0.743351   
n02085620-Chihuahua\n02085620_13151.jpg             0.117868   0.257813   
...                                                      ...        ...   
n02116738-African_hunting_dog\n02116738_10614.jpg   0.144953   1.881053   
n02116738-African_hunting_dog\n02116738_9282.jpg    2.168356   2.056204   
n02116738-African_hunting_dog\n02116738_6754.jpg    0.442680   2.221987   
n02116738-African_hunting_dog\n02116738_9333.jpg    0.421884   0.610241   
n02116738-African_hunting_dog\n02116738_2503.jpg    0.000000   1.390403   

                                                   feature_6  feature_7  \
image_path                                                                
n02085620-Chihuahua\n02085620_5927.jpg              0.073741   0.260721   
n02085620-Chihuahua\n02085620_4441.jpg              0.352944   0.616292   
n02085620-Chihuahua\n02085620_1502.jpg              0.772889   0.267891   
n02085620-Chihuahua\n02085620_1916.jpg              0.053917   0.293086   
n02085620-Chihuahua\n02085620_13151.jpg             0.817096   0.777969   
...                                                      ...        ...   
n02116738-African_hunting_dog\n02116738_10614.jpg   1.025135   1.309460   
n02116738-African_hunting_dog\n02116738_9282.jpg    0.418436   0.592864   
n02116738-African_hunting_dog\n02116738_6754.jpg    1.028761   2.321181   
n02116738-African_hunting_dog\n02116738_9333.jpg    0.311903   1.028679   
n02116738-African_hunting_dog\n02116738_2503.jpg    0.759935   2.652178   

                                                   feature_8  feature_9  ...  \
image_path                                                               ...   
n02085620-Chihuahua\n02085

In [11]:
df_test = reformat_df(df_test)
df_test

target  feature_1  \
image_path                                                            
n02085620-Chihuahua\n02085620_2650.jpg                 1   2.750876   
n02085620-Chihuahua\n02085620_4919.jpg                 1   3.723587   
n02085620-Chihuahua\n02085620_1765.jpg                 1   0.971007   
n02085620-Chihuahua\n02085620_3006.jpg                 1   1.874442   
n02085620-Chihuahua\n02085620_1492.jpg                 1   0.254995   
...                                                  ...        ...   
n02116738-African_hunting_dog\n02116738_4991.jpg     120   0.068341   
n02116738-African_hunting_dog\n02116738_3024.jpg     120   0.012424   
n02116738-African_hunting_dog\n02116738_3635.jpg     120   0.090567   
n02116738-African_hunting_dog\n02116738_2988.jpg     120   0.249121   
n02116738-African_hunting_dog\n02116738_6330.jpg     120   0.006490   

                                                  feature_2  feature_3  \
image_path                                                               
n02085620-Chihuahua\n02085620_2650.jpg             0.402823   0.490048   
n02085620-Chihuahua\n02085620_4919.jpg             1.346295   0.413855   
n02085620-Chihuahua\n02085620_1765.jpg             0.447430   0.136485   
n02085620-Chihuahua\n02085620_3006.jpg             1.340326   0.430788   
n02085620-Chihuahua\n02085620_1492.jpg             1.975686   0.795596   
...                                                     ...        ...   
n02116738-African_hunting_dog\n02116738_4991.jpg   0.824908   0.193437   
n02116738-African_hunting_dog\n02116738_3024.jpg   0.074204   1.992666   
n02116738-African_hunting_dog\n02116738_3635.jpg   0.394184   1.207222   
n02116738-African_hunting_dog\n02116738_2988.jpg   0.176359   2.857338   
n02116738-African_hunting_dog\n02116738_6330.jpg   0.197918   0.512689   

                                                  feature_4  feature_5  \
image_path                                                               
n02085620-Chihuahua\n02085620_2650.jpg             0.088944   0.067087   
n02085620-Chihuahua\n02085620_4919.jpg             1.559932   2.922223   
n02085620-Chihuahua\n02085620_1765.jpg             0.470510   0.371954   
n02085620-Chihuahua\n02085620_3006.jpg             1.061602   0.018172   
n02085620-Chihuahua\n02085620_1492.jpg             0.792883   0.491056   
...                                                     ...        ...   
n02116738-African_hunting_dog\n02116738_4991.jpg   0.906468   2.191380   
n02116738-African_hunting_dog\n02116738_3024.jpg   0.076113   0.573071   
n02116738-African_hunting_dog\n02116738_3635.jpg   0.123069   2.618800   
n02116738-African_hunting_dog\n02116738_2988.jpg   0.408916   0.582186   
n02116738-African_hunting_dog\n02116738_6330.jpg   0.059964   0.486442   

                                                  feature_6  feature_7  \
image_path                                                               
n02085620-Chihuahua\n02085620_2650.jpg             0.097192   0.188324   
n02085620-Chihuahua\n02085620_4919.jpg             0.046513   0.065089   
n02085620-Chihuahua\n02085620_1765.jpg             0.112476   0.265642   
n02085620-Chihuahua\n02085620_3006.jpg             0.659020   0.024218   
n02085620-Chihuahua\n02085620_1492.jpg             1.389314   0.066709   
...                                                     ...        ...   
n02116738-African_hunting_dog\n02116738_4991.jpg   0.567653   2.182863   
n02116738-African_hunting_dog\n02116738_3024.jpg   1.246892   0.586113   
n02116738-African_hunting_dog\n02116738_3635.jpg   0.823760   3.099759   
n02116738-African_hunting_dog\n02116738_2988.jpg   1.393884   2.366030   
n02116738-African_hunting_dog\n02116738_6330.jpg   0.658949   3.135226   

                                                  feature_8  feature_9  ...  \
image_path                                                              ...   
n02085620-Chihuahua\n02085620_2650.jpg             0.554036   0.597607  ...   


# **Create train/validation/test splits**
* From the training set, we will create a small validation set which is going to be used to find the best Hyperparameter for the models.
* We cant use the test set because that will lead to something called model leakage!!!

In [12]:
# train features
X = df_train.iloc[:,1:]
# train labels
y = df_train.iloc[:,0]

# create the training & validation sets
X_train, X_valid, y_train, y_valid = train_test_split(X, y,test_size=0.15, shuffle=True, random_state=7)

# re-assign the test set variables for consistency & shuffle
df_test = df_test.sample(frac = 1)
X_test = df_test.iloc[:,1:]
y_test = df_test.iloc[:,0]

In [13]:
print(f"Training Data: {X_train.shape}")
print(f"Training Labels: {y_train.shape}")

print(f"Validation Data: {X_valid.shape}")
print(f"Validation Labels: {y_valid.shape}")

print(f"Testing Data: {X_test.shape}")
print(f"Testing Labels: {y_test.shape}")

Training Data: (10200, 512)
Training Labels: (10200,)
Validation Data: (1800, 512)
Validation Labels: (1800,)
Testing Data: (8580, 512)
Testing Labels: (8580,)


# **Preprocessing**
* Training & Validation datasets will be Normalised
* Most ML models expect the class values to start from zero to n. Our targets start from one to n, hence we need to fix this up.
* PCA will be used to create a secondary dataset which you will use to experiment to see if reducing the attributes will lead to better classification results

## 1) **Normalize values**


In [14]:
# initialize the StandardScaler
stdsclr = StandardScaler()
# fit it to the trainset ONLY
stdsclr.fit(X_train)
# Apply it to Train data
X_train_sclr = stdsclr.transform(X_train)
# Apply it to Valid data
X_valid_sclr = stdsclr.transform(X_valid)

In [15]:
print(f"Train Data Scaled: {X_train_sclr.shape}")
print(f"Validation Data Scaled: {X_valid_sclr.shape}")

Train Data Scaled: (10200, 512)
Validation Data Scaled: (1800, 512)


## **2) Rebase target classes**
* subtract 1 from all target labels to shift the range from 0 to 119 instead of 1 to 120

In [16]:
y_train = y_train - 1
y_valid = y_valid - 1
y_test = y_test - 1

In [17]:
y_train.value_counts().sort_index()

target
0      84
1      84
2      93
3      89
4      90
       ..
115    84
116    85
117    86
118    82
119    79
Name: count, Length: 120, dtype: int64

## **3) PCA**
* A secondary dataset used for the experiment

In [18]:
# create PCA to keep 95% of the variance
pca = PCA(n_components=0.95)

# Fit PCA on the training dataset
pca.fit(X_train_sclr)

# Transform the training, validation, and test datasets
X_train_pca = pca.transform(X_train_sclr)
X_valid_pca = pca.transform(X_valid_sclr)
X_test_pca = pca.transform(X_test)

# Output the results
print("Transformed training set shape:", X_train_pca.shape)
print("Transformed validation set shape:", X_valid_pca.shape)
print("Transformed test set shape:", X_test_pca.shape)

Transformed training set shape: (10200, 256)
Transformed validation set shape: (1800, 256)
Transformed test set shape: (8580, 256)


/opt/conda/lib/python3.10/site-packages/sklearn/base.py:432: UserWarning: X has feature names, but PCA was fitted without feature names
  warnings.warn(


In [ ]:
print(f"Training Data PCA: {X_train_pca.shape}")
print(f"Training Labels: {y_train.shape}")

print(f"Validation Data PCA: {X_valid_pca.shape}")
print(f"Validation Labels: {y_valid.shape}")

print(f"Testing Data PCA: {X_test_pca.shape}")
print(f"Testing Labels: {y_test.shape}")

Training Data PCA: (10200, 256)

Training Labels: (10200,)

Validation Data PCA: (1800, 256)

Validation Labels: (1800,)

Testing Data PCA: (8580, 256)

Testing Labels: (8580,)


# **Display Data Variables**

## **Dataset 1 - all features**

In [ ]:
# train
X_train_sclr
y_train
# validation
X_valid_sclr
y_valid
# test
X_test
y_test

image_path
n02106550-Rottweiler\n02106550_7419.jpg                         83
n02112350-keeshond\n02112350_3955.jpg                          109
n02101388-Brittany_spaniel\n02101388_2076.jpg                   64
n02094433-Yorkshire_terrier\n02094433_2824.jpg                  36
n02111277-Newfoundland\n02111277_3741.jpg                      104
                                                              ... 
n02095570-Lakeland_terrier\n02095570_375.jpg                    38
n02093428-American_Staffordshire_terrier\n02093428_1767.jpg     29
n02110185-Siberian_husky\n02110185_9833.jpg                     99
n02107908-Appenzeller\n02107908_6567.jpg                        89
n02111500-Great_Pyrenees\n02111500_3548.jpg                    105
Name: target, Length: 8580, dtype: int64

## **Dataset 2 - PCA**

In [ ]:
# Dataset 2 - PCA
# train
X_train_pca
y_train
# validation
X_valid_pca
y_valid
# test
X_test_pca
y_test

image_path
n02106550-Rottweiler\n02106550_7419.jpg                         83
n02112350-keeshond\n02112350_3955.jpg                          109
n02101388-Brittany_spaniel\n02101388_2076.jpg                   64
n02094433-Yorkshire_terrier\n02094433_2824.jpg                  36
n02111277-Newfoundland\n02111277_3741.jpg                      104
                                                              ... 
n02095570-Lakeland_terrier\n02095570_375.jpg                    38
n02093428-American_Staffordshire_terrier\n02093428_1767.jpg     29
n02110185-Siberian_husky\n02110185_9833.jpg                     99
n02107908-Appenzeller\n02107908_6567.jpg                        89
n02111500-Great_Pyrenees\n02111500_3548.jpg                    105
Name: target, Length: 8580, dtype: int64

# **Model 1**

In [ ]:
# Example Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics

## **1) run model with default parameters**

In [ ]:
# create the model
gnb = GaussianNB()

# train the model
gnb.fit(X_train_sclr, y_train)

# Predict on the test set
y_pred = gnb.predict(X_test)

print("Accuracy:", metrics.accuracy_score(y_test, y_pred))

/usr/local/lib/python3.10/dist-packages/sklearn/base.py:432: UserWarning: X has feature names, but GaussianNB was fitted without feature names

  warnings.warn(


Accuracy: 0.42925407925407927


## **2) Conduct Hyperparameter tuning**

## **3) Compare Pre & Post Results**

## 4) **Repeat the above steps but for the PCA dataset**

# **Model 2**

# **Model 3**